### PR TITLE
Fix some bugs in LLVMTypeHierarchy.cpp

### PR DESCRIPTION
**Pull request by benquike**
in methods like removeStructOrClassPrefix, ltrim is used to remove a prefix
from a name, which may be wrong. According to the comments of ltrim, ltrim
removes all chars belonging to the prefix from the target string, instead
of removing the prefix directly. The right way is simply removing the prefix
using drop_front

### CODE REVIEW

**Review by vulder — CHANGES_REQUESTED** [review]
`ltrim` only drops consecutive chars from left to right, so it should not be a problem in general, however, replacing it with a direct call to `drop_front` seems fine after we already checked that the string starts with the correct substring. (Info: `ltrim` is implemented on top of `drop_front`)
Except for my to comments the tests seem fine to me.

But @pdschubert will double check anyhow :smile:

File: lib/PhasarLLVM/TypeHierarchy/LLVMTypeHierarchy.cpp [hunk 1]
```diff
@@ -96,32 +96,32 @@ std::string
...
-    return SR.ltrim(ClassPrefix);
+    return SR.drop_front(ClassPrefix.size());
   }
   return TypeName;
 }
 
 std::string LLVMTypeHierarchy::removeTypeInfoPrefix(std::string VarName) {
   llvm::StringRef SR(VarName);
   if (SR.startswith(TypeInfoPrefixDemang)) {
-    return SR.ltrim(TypeInfoPrefixDemang);
```
Comment (benquike) [inline lib/PhasarLLVM/TypeHierarchy/LLVMTypeHierarchy.cpp:110]: Here `ltrim` is defined as 

```
 LLVM_NODISCARD
     StringRef ltrim(StringRef Chars = " \t\n\v\f\r") const {
       return drop_front(std::min(Length, find_first_not_of(Chars)));
     }
```

Which means that any prefix containing chars in `Chars` will be removed,
If all the chars in SR are included in TypeInfoPrefixDemang, the returned value will be an empty string.
E.g. given, TypeInfoPrefixDemang = "typeinfo for ", and SR = "typeinfo for yointtttttt", the returned value would be empty.

Here what we need is simply remove the the prefix (Chars).

File: lib/PhasarLLVM/TypeHierarchy/LLVMTypeHierarchy.cpp [hunk 2]
```diff
@@ -220,6 +242,7 @@ LLVMTypeHierarchy::getVirtualFunctions(const llvm::Module &M,
       }
     }
   }
+ret:
```
[the remark this answers is not in the document]
Comment (benquike) [reply lib/PhasarLLVM/TypeHierarchy/LLVMTypeHierarchy.cpp]: updated

File: lib/PhasarLLVM/TypeHierarchy/LLVMTypeHierarchy.cpp [hunk 3]
```diff
@@ -197,6 +206,13 @@ LLVMTypeHierarchy::getVirtualFunctions(const llvm::Module &M,
   std::vector<const llvm::Function *> VFS;
   if (const auto *TV = ClearNameTVMap[ClearName]) {
     if (const auto *TI = llvm::dyn_cast<llvm::GlobalVariable>(TV)) {
+
+      if (!TI->hasInitializer()) {
+        LOG_IF_ENABLE(BOOST_LOG_SEV(lg::get(), DEBUG)
+                      << ClearName << " does not have initializer");
+        goto ret;
```
[the remark this answers is not in the document]
Comment (benquike) [reply lib/PhasarLLVM/TypeHierarchy/LLVMTypeHierarchy.cpp]: udpated

File: lib/PhasarLLVM/TypeHierarchy/LLVMTypeHierarchy.cpp [hunk 4]
```diff
@@ -163,7 +163,15 @@ LLVMTypeHierarchy::getSubTypes(const llvm::Module &M,
...
   // find corresponding type info variable
   std::vector<const llvm::StructType *> SubTypes;
-  if (const auto *TI = ClearNameTIMap[removeStructOrClassPrefix(Type)]) {
+  std::string ClearName = removeStructOrClassPrefix(Type);
+  if (const auto *TI = ClearNameTIMap[ClearName]) {
+
+    if (!TI->hasInitializer()) {
+      LOG_IF_ENABLE(BOOST_LOG_SEV(lg::get(), DEBUG)
+                    << ClearName << " does not have initializer");
+      goto ret;
```
[the remark this answers is not in the document]
Comment (benquike) [reply lib/PhasarLLVM/TypeHierarchy/LLVMTypeHierarchy.cpp]: updated

File: lib/PhasarLLVM/TypeHierarchy/LLVMTypeHierarchy.cpp (hunk 1 above)
Comment (vulder) [reply lib/PhasarLLVM/TypeHierarchy/LLVMTypeHierarchy.cpp:110]: A, I see what you mean.

**Review by vulder — APPROVED** [review]
lgtm

**Review by pdschubert — APPROVED** [review]
Thank you very much!

**Pull request merged**